File: electron/build-backend.js
const { spawnSync } = require("node:child_process");
const fs = require("node:fs");
const path = require("node:path");

const root = path.resolve(__dirname, "..");

function getPythonCommand() {
  if (process.env.DK_PYTHON) {
    return process.env.DK_PYTHON;
  }

  const venvPython = process.platform === "win32"
    ? path.join(root, "venv", "Scripts", "python.exe")
    : path.join(root, "venv", "bin", "python");

  if (fs.existsSync(venvPython)) {
    return venvPython;
  }

  return process.platform === "win32" ? "python" : "python3";
}

const python = getPythonCommand();
const result = spawnSync(
  python,
  ["-m", "PyInstaller", "--clean", "--noconfirm", "electron/backend.spec"],
  {
    cwd: root,
    env: {
      ...process.env,
      PYTHONPATH: root
    },
    stdio: "inherit"
  }
);

if (result.error) {
  console.error(result.error.message);
  process.exit(1);
}

if (result.status !== 0) {
  console.error(
    "Backend build failed. Install build dependencies with: python -m pip install -r requirements-build.txt"
  );
  process.exit(result.status || 1);
}

File: electron/build-backend.py
import os
import subprocess
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]


def get_python_command() -> str:
    if os.environ.get("DK_PYTHON"):
        return os.environ["DK_PYTHON"]

    venv_python = (
        ROOT / "venv" / "Scripts" / "python.exe"
        if sys.platform == "win32"
        else ROOT / "venv" / "bin" / "python"
    )

    if venv_python.exists():
        return str(venv_python)

    return sys.executable


def main() -> int:
    python = get_python_command()
    env = os.environ.copy()
    env["PYTHONPATH"] = str(ROOT)

    try:
        result = subprocess.run(
            [
                python,
                "-m",
                "PyInstaller",
                "--clean",
                "--noconfirm",
                "electron/backend.spec",
            ],
            cwd=ROOT,
            env=env,
        )

    except OSError as error:
        print(error, file=sys.stderr)
        return 1

    if result.returncode != 0:
        print(
            "Backend build failed. Install build dependencies with: "
            "python -m pip install -r requirements-build.txt",
            file=sys.stderr,
        )
        return result.returncode or 1

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
